File: main/get_expected_score.m
function [EH, HE] = get_expected_score(S, mu, sigma, ...
    noise_factor, scale, query_selected, num_particles)
% GET_EXPECTED_SCORE Computes the terms of the active learning heuristic:
% the difference between the entropy of the expected label and the expected 
% entropy of the label given the classifier
% 
% Input arguments:
%  S                - Embeddings of words in set.
%  mu               - Mean of the classifier distribution
%  sigma            - Covariance matrix of the classifier distribution
%  noise_factor     - Scaling factor to control noise level model of the labeling
%  scale            - Scaling factor to control noise level in the model of the example selection
%  query_selected   - 1: q_pos or 2: q_neg
%  num_particles    - Number of particles used for the approximation (optional)
%
% Returns:
%  EH - The expected entropy of the signal (E(H(Y)))
%  HE - The entropy of the expected signal (H(Expectation(Y)))

    %  If num_particles is not given, set a default value
    if nargin < 9
        num_particles = 1e4;
    end
    %  If query is not given, set q_pos
    if nargin < 8
        query_selected = 1;
    end

    % Define likelihood functions modeling the human response
    p_y1_xandtheta = @(theta, S) 1 ./ (1 + exp(- noise_factor * (theta' * S))); 
    if query_selected == 1
        softmax_num = @(theta, S) exp((scale * (theta' * S)));
    else
        softmax_num = @(theta, S) exp(-(scale * (theta' * S)));
    end

    % Sample and normalize particles
    particles = mvnrnd(mu, sigma, num_particles)';
    particles = particles ./ vecnorm(particles);

    % Compute H(Expectation(Y)) and E(H(Y))
    % py1_all is a N_particles x N_S*2
    % -> *2 for y=1 and y=-1 options -> N_S*2 total combination of x and y
    p_y1_words = p_y1_xandtheta(particles, S);
    p_y_all = [p_y1_words, 1-p_y1_words];

    % Computejoint discrete pmf f(x, y)
    softmax_num_words = softmax_num(particles, S);
    softmax_den_words = sum(softmax_num_words, 2);
    p_x_particles_all = softmax_num_words ./ softmax_den_words;
    p_xy_particles_all= p_y_all .* repmat(p_x_particles_all, [1, 2]);

    % Compute the Entropy Scores
    xylogxy_all = sum(p_xy_particles_all .* log2(p_xy_particles_all), 2);
    EH = mean(-xylogxy_all);
    E_p_xy = mean(p_xy_particles_all);
    HE = -sum(E_p_xy .* log2(E_p_xy));
end

File: python/get_expected_score.py
import numpy as np

def get_expected_score(S, mu, sigma, noise_factor, scale, query_selected=1, num_particles=1e4):
    """
    Computes the terms of the active learning heuristic: 
    the difference between the entropy of the expected label and the expected 
    entropy of the label given the classifier.
    
    Parameters:
    S              - Embeddings of words in the set (matrix).
    mu             - Mean of the classifier distribution (vector).
    sigma          - Covariance matrix of the classifier distribution (matrix).
    noise_factor   - Scaling factor to control noise level model of the labeling.
    scale          - Scaling factor to control noise level in the model of the example selection.
    query_selected - 1: q_pos or 2: q_neg (default: 1).
    num_particles  - Number of particles used for the approximation (default: 1e4).
    
    Returns:
    EH - The expected entropy of the signal (E(H(Y))).
    HE - The entropy of the expected signal (H(Expectation(Y))).
    """
    
    # Define likelihood functions modeling the human response
    def p_y1_xandtheta(theta, S):
        return 1.0 / (1.0 + np.exp(- noise_factor * (theta.T @ S)))

    if query_selected == 1:
        softmax_num = lambda theta, S: np.exp(scale * (theta.T @ S))
    else:
        softmax_num = lambda theta, S: np.exp(-scale * (theta.T @ S))

    # Sample and normalize particles
    particles = np.random.multivariate_normal(mu, sigma, num_particles).T
    particles = particles / np.linalg.norm(particles, axis=0)  # Normalize each particle

    # Compute p(Y=1|x,theta) for all particles and embeddings
    p_y1_words = p_y1_xandtheta(particles, S)
    p_y_all = np.hstack([p_y1_words, 1 - p_y1_words])  # Combine for y=1 and y=-1

    # Compute joint discrete pmf f(x, y)
    softmax_num_words = softmax_num(particles, S)
    softmax_den_words = np.sum(softmax_num_words, axis=1)
    softmax_den_words = softmax_den_words.reshape(-1, 1) # OK until this point
    p_x_particles_all = softmax_num_words / softmax_den_words
    p_xy_particles_all = p_y_all * np.tile(p_x_particles_all, (1, 2))

    # Compute the entropy scores
    xylogxy_all = np.sum(p_xy_particles_all * np.log2(p_xy_particles_all), axis=1) 
    EH = np.mean(-xylogxy_all)

    E_p_xy = np.mean(p_xy_particles_all, axis=0)
    HE = -np.sum(E_p_xy * np.log2(E_p_xy))

    return EH, HE
